src/deepsparse/server/openai_server.py

```diff
@@ -0,0 +1,385 @@
+
+OPENAI_CHAT_NOT_SUPPORTED = ["logit_bias", "best_ok", "ignore_eos", "use_beam_search"]
+OPENAI_TO_DEEPSPARSE_MAPPINGS = {
+    "max_tokens": "max_length",
+    "frequency_penalty": "repetition_penalty",
+}
+SUPPORTED_TASKS = ["text_generation", "opt", "bloom"]
+
+
+class OpenAIServer(Server):
+    def _add_routes(self, app: FastAPI):
+        for endpoint_config in self.server_config.endpoints:
+            self._add_endpoint(
+                app,
+                endpoint_config,
+            )
+
+        _LOGGER.info(f"Added endpoints: {[route.path for route in app.routes]}")
+        return app
+
+    def _add_endpoint(
+        self,
+        app: FastAPI,
+        endpoint_config: EndpointConfig,
+    ):
+        pipeline_config = endpoint_config.to_pipeline_config()
+        pipeline_config.kwargs["executor"] = self.executor
+
+        _LOGGER.info(f"Initializing pipeline for '{endpoint_config.name}'")
+
+        if pipeline_config.task not in SUPPORTED_TASKS:
+            raise ValueError(
+                "OpenAI API is only available for one of the following "
+                f"tasks: {SUPPORTED_TASKS}"
+            )
+
+        pipeline = Pipeline.from_config(
+            pipeline_config, self.context, self.server_logger
+        )
+
+        _LOGGER.info(f"Adding endpoints for '{endpoint_config.name}'")
+        self._add_chat_completion_endpoint(
+            app,
+            endpoint_config,
+            pipeline,
+        )
+
+    def _add_chat_completion_endpoint(
+        self,
+        app: FastAPI,
+        endpoint_config: EndpointConfig,
+        pipeline: Pipeline,
+    ):
+        routes_and_fns = []
+        route = (
+            f"{endpoint_config.route}/chat/completions"
+            if endpoint_config.route
+            else f"/v2/models/{endpoint_config.name}/chat/completions"
+        )
+        route = self.clean_up_route(route)
+        routes_and_fns.append(
+            (
+                route,
+                partial(OpenAIServer.create_chat_completion, ProxyPipeline(pipeline)),
+            )
+        )
+
+        self._update_routes(
+            app=app,
+            routes_and_fns=routes_and_fns,
+            response_model=ChatCompletionResponse,
+            methods=["POST"],
+            tags=["model", "inference"],
+        )
+
+    @staticmethod
+    async def generate(
+        prompt: str, request_id: str, generation_kwargs: dict, pipeline: Pipeline
+    ) -> AsyncGenerator[RequestOutput, None]:
+        def tokenize(text: str) -> List[int]:
+            return pipeline.tokenizer(text)
+
+        prompt_token_ids = tokenize(prompt)
+        generation_kwargs = map_generation_schema(generation_kwargs)
+
+        stream = generation_kwargs["stream"]
+        presence_penalty = generation_kwargs["presence_penalty"]
+        stop = generation_kwargs["stop"]
+
+        generation_kwargs.pop("stream")
+        generation_kwargs.pop("presence_penalty")
+        generation_kwargs.pop("stop")
```

Comment:
are we guaranteed to have these keys? additionally, pop should return the value, so we should be able to squash these lines, no?

Reply:
Yup, the request is mapped to the `ChatCompletionRequest` schema, which contains all these fields so will have defaults, even if the user does not define them in their request.  And yup, squashing. Good catch.